### PR TITLE
DDP: 10% of NCCL backend perf improvements with mixed-prec support

### PR DESCRIPTION
**Pull request by teng-li**
This PR lets NCCL backend to directly enqueue the NCCL reduction kernels in the same thread since it's an async call by nature, and everything now goes to the default stream. This essentially gets rid of the overheads of python thread synchronization, stream synchronization, as well as bucketing map lookup overheads.  

For DDP with multiple GPU (the default use case), instead of doing a two step reduction, we use the new NCCL backend API to all-reduce all GPU at one time, this is basically 4 times all-reduce throughput on DGX1s compared to the all-reduce(single GPU version) plus the intra-node reduce overheads.

As a results, we have the following perf improvements.

For DDP with 8 GPU (default use case):

On two DGX1s with 8 V100s, 256 batch size / process, single process / node
ResNet50
**0.139 sec / iter** reduced from **0.154 sec / iter** (about 10 percent improvements)
ResNet101
**0.247 sec / iter** reduced from **0.272 sec / iter** (about 10 percent improvements)

For DDP with 1 GPU (multi-process use case)

On singe DGX1s with 8 V100s, ResNet50, 32 batch size per GPU and process, 8 processes distributed training 
**0.109 sec / iter** reduced from **0.116 sec / iter** (about 6 percent improvement)

In addition,  added bucketing to limit the memory usage, added mixed precision support.

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -222,6 +240,66 @@ def _register_grad_hooks(self):
...
+            # (2) inter-node allreduce for all the first lead GPUs in all nodes
+            dist.all_reduce_multigpu(all_grads_coalesced,
+                                     group=self.nccl_reduction_group_id)
+
+            # Now only work on the first lead GPU
+            all_grads_coalesced[0] /= dist.get_world_size()
+            for grad, reduced in \
+                    zip(all_grads[0],
+                        _unflatten_dense_tensors(all_grads_coalesced[0],
+                                                 all_grads[0])):
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: ouch. Let's do something about this line, because it's impossible to read. How about this:
```python
iter = zip(...)
for grad, reduced in iter:
    ...
```

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -222,6 +240,66 @@ def _register_grad_hooks(self):
...
+        def reduction_fn_nccl():
+            # This function only needs to be called once
+            if not self.need_reduction:
+                return
+            self.need_reduction = False
+            all_grads = [[] for _ in range(len(self._module_copies))]
+            all_grads_coalesced = []
+
+            # Coalesce all the gradients
+            # TODO: Add mixed precision support here
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: `_take_tensors` is your friend again

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -222,6 +240,66 @@ def _register_grad_hooks(self):
...
+                    if not param.requires_grad or param.grad is None:
+                        continue
+                    if param.grad.requires_grad:
+                        raise RuntimeError("DistributedDataParallel only works "
+                                           "with gradients that don't require "
+                                           "grad")
+                    # Adding the gradients for reduction
+                    all_grads[idx].append(param.grad.data)
+                with torch.cuda.device(self.device_ids[idx]):
+                    dev_grads_coalesced = _flatten_dense_tensors(all_grads[idx])
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: This has no bound on the size, so you'll possibly double the memory usage. Use `_take_tensors` to bucket them appropriately.

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -222,6 +240,66 @@ def _register_grad_hooks(self):
...
+            dist.all_reduce_multigpu(all_grads_coalesced,
+                                     group=self.nccl_reduction_group_id)
+
+            # Now only work on the first lead GPU
+            all_grads_coalesced[0] /= dist.get_world_size()
+            for grad, reduced in \
+                    zip(all_grads[0],
+                        _unflatten_dense_tensors(all_grads_coalesced[0],
+                                                 all_grads[0])):
+                grad.copy_(reduced)
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: If the grads don't require grad it's faster to do `grad.data.set_(reduced)`

File: torch/nn/parallel/distributed.py [hunk 5]
```diff
@@ -222,6 +240,66 @@ def _register_grad_hooks(self):
...
+
+        # Now register the reduction function in the execution engine
+        for module in self._module_copies:
+            for p in module.parameters():
+                if p.requires_grad:
+                    def allreduce_hook(*unused):
+                        Variable._execution_engine.\
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Can you please not fold the line in here? We allow up to 120 chars, so it should fit. Alternatively, to reduce the nesting level, do: `if not p.requires_grad: continue` above, and dedent this

File: torch/nn/parallel/distributed.py [hunk 6]
```diff
@@ -222,6 +240,66 @@ def _register_grad_hooks(self):
...
+            # This single op will do all-reduce on all GPUs utilizing multiple
+            # all reduce rings when we have more than one fast IB interfaces.
+            # We will only use device 0's results, but this single op should be
+            # faster than doing the following two operation sequentially:
+            # (1) intra-node reduce to lead GPU, followed by
+            # (2) inter-node allreduce for all the first lead GPUs in all nodes
+            dist.all_reduce_multigpu(all_grads_coalesced,
+                                     group=self.nccl_reduction_group_id)
+
+            # Now only work on the first lead GPU
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: it's not about working on the "first" GPU, it's more about working on the original module (there's an invariant that the first element of `_module_copies` is what's seen outside

Comment (teng-li) [reply torch/nn/parallel/distributed.py]: Yeah, I meant that :D

File: torch/nn/parallel/distributed.py (hunk 5 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: 80 line limit is FB's coding style :) So I think a lot of the PR does that too.

File: torch/nn/parallel/distributed.py [hunk 7]
```diff
@@ -222,7 +243,85 @@ def _register_grad_hooks(self):
...
+
+                # Now bucketing the parameters
+                dev_grads_buckets = _take_tensors(all_grads[dev_idx],
+                                                  self.nccl_reduce_bucket_size)
+                if dev_idx == 0:
+                    mst_dev_grads_buckets = list(dev_grads_buckets)
+                    all_grads_coalesced = \
+                        [[] for _ in range(len(mst_dev_grads_buckets))]
+
+                for bkt_idx, dev_grads in enumerate(dev_grads_buckets):
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Is it expected that this loop will execute if and only if `dev_idx != 0`? `take_tensors` returns an iterator and doing `list(dev_grads_buckets)` will deplete it, meaning that this will terminate immediately. If this is on purpose, please add this loop in an `else` branch, to make it clear that it was your intent.

File: torch/nn/parallel/distributed.py [hunk 8]
```diff
@@ -222,7 +243,85 @@ def _register_grad_hooks(self):
...
+                    # Coalescing the bucket
+                    dev_id = self.device_ids[dev_idx]
+                    with torch.cuda.device(dev_id):
+                        dev_grads_coalesced = _flatten_dense_tensors(dev_grads)
+                        all_grads_coalesced[bkt_idx].append(dev_grads_coalesced)
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Well, flattening them all at one time, before you even start broadcasting defeats the whole purpose of bucketing, since you will still use 2x the memory needed for parameters.

File: torch/nn/parallel/distributed.py [hunk 9]
```diff
@@ -222,7 +243,85 @@ def _register_grad_hooks(self):
...
+
+            # Reduce all the gradients first
+            # This single op will do all-reduce on all GPUs utilizing multiple
+            # all reduce rings when we have more than one fast IB interfaces.
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: These 3 lines are unnecessary. It's clear that you're reducing the gradients from the code, and you don't need to provide the docs for `all_reduce_multigpu` in here (+ it doesn't even apply to the environment many users work in).

**Review by apaszke — COMMENTED** [review]
Just out of curiosity, have you tested this using NCCL2, but *without IB*? It might no longer be the fastest way to do things if that's the environment someone is in, and only few people have IB.

File: torch/nn/parallel/distributed.py (hunk 8 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: We are flattening one bucket at one time here. Why do you think it's all at one time?

File: torch/nn/parallel/distributed.py (hunk 7 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: This loop is expected to execute on all dev_idx (ex)

File: torch/nn/parallel/distributed.py (hunk 9 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: Ok

File: torch/nn/parallel/distributed.py [hunk 10]
```diff
@@ -222,7 +243,85 @@ def _register_grad_hooks(self):
...
+            # the gradients for each bucket
+            for grads_coalesced, mst_dev_grads in zip(all_grads_coalesced,
+                                                      mst_dev_grads_buckets):
+                grads_coalesced[0] /= dist.get_world_size()
+                grads_reduced = _unflatten_dense_tensors(grads_coalesced[0],
+                                                         mst_dev_grads)
+                for grad, reduced in zip(mst_dev_grads, grads_reduced):
+                    grad.copy_(reduced)
+
+        # Now register the reduction function in the execution engine
```
[the remark this answers is not in the document]
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: I meant that, yeah, will change the comments

File: torch/nn/parallel/distributed.py [hunk 11]
```diff
@@ -222,7 +243,85 @@ def _register_grad_hooks(self):
...
+        for module in self._module_copies:
```
[the remark this answers is not in the document]
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: Yeah, I think so too

**Review by teng-li — COMMENTED** [review]
@apaszke  I haven't got a chance to test on Ethernet yet. But for Nccl backend, this should be the same, since the other code path, we are also using a single bucket reduction, without any threading overlapping

File: torch/nn/parallel/distributed.py (hunk 7 above)
Comment (apaszke) [reply torch/nn/parallel/distributed.py]: Then I think it won't run on `dev_idx == 0`

File: torch/nn/parallel/distributed.py (hunk 8 above)
Comment (apaszke) [reply torch/nn/parallel/distributed.py]: You are flattening them one at a time, but you are first flattening everything and storing the buffers. Basically you're doing this:
```python
flattened = [flatten(bucket) for bucket in buckets] # NB: 2x memory used now
for flat in flattened:
    ... # process flat tensors
```
instead of this:
```python
for bucket in buckets:
    flat = flatten(bucket) # NOTE: flattening inside the loop - is freed at next iteration
    ... # process flat tensor
```

**Comment by apaszke** [comment]
I'm sorry, why should it be the same? Ethernet is likely to have much lower throughput than 4xIB that you tested with, and interleaving the communication with backward might still be beneficial in that case.

**Comment by teng-li** [comment]
@apaszke I meant, we don't currently support multi-threading for NCCL. Remember that the gradient buckets of each process on different nodes need to be executed in the exact same order. So even if we use the old code path, we still limit the number of bucket to be 1.

I will further test the Ethernet perf, which is on my to-do list anyway.  But this should block this PR for now

**Comment by apaszke** [comment]
Is this true, even if they are using different groups? I thought that it doesn't matter as long as they use different communicators, right?

**Comment by teng-li** [comment]
@apaszke right, they are using different communicators.  But the NCCL call order (among all the reduction thread) needs to be maintained among all the nodes. The gradients available order for each bucket can not be guaranteed for different nodes.

**Comment by apaszke** [comment]
Does it have to be the same, even when they are on different communicators? I think the whole purpose of keeping multiple comms was to allow concurrent operations to execute independently

**Comment by teng-li** [comment]
@apaszke  according to @csarofeen, the order needs to be maintained.

**Comment by csarofeen** [comment]
I believe if your communicators overlap, you can get in trouble if the order is not correct. I don't believe you can have 2 communicators on 2 gpu's and have:
GPU 0 call on comm 0 then comm 1
and
GPU 1 call on comm 1 then comm 0

**Comment by apaszke** [comment]
Aren't these situations the reason why `ncclGroupStart` and `ncclGroupEnd` were introduced?

**Comment by teng-li** [comment]
@apaszke  comments addressed

File: torch/nn/parallel/distributed.py [hunk 12]
```diff
@@ -222,7 +243,79 @@ def _register_grad_hooks(self):
...
+                    with torch.cuda.device(dev_id):
+                        dev_grads_batch_coalesced = _flatten_dense_tensors(dev_grads_batch)
+                        grads_batch_coalesced.append(dev_grads_batch_coalesced)
+
+                # We will only use device 0's results, but this single op should be
+                # faster than doing the following two operation sequentially:
+                # (1) intra-node reduce to lead GPU, followed by
+                # (2) inter-node allreduce for all the first lead GPUs in all nodes
+                dist.all_reduce_multigpu(grads_batch_coalesced,
+                                         group=self.nccl_reduction_group_id)
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Do we ever zero the gradients on the replicas? It's quite important as they will start to accumulate and will be included in this sum

**Comment by ngimel** [comment]
@apaszke ncclGroupStart and ncclGroupEnd were introduced for a case when a single thread submitted nccl calls for the different GPUs (the way DataParallel operates). In nccl2 it would deadlock unless wrapped in groupStart/groupEnd. For a single GPU per process groupStart/groupEnd is a noop.

**Comment by ngimel** [comment]
Got answer from nccl: "calling collectives on different comms is to be avoided, and if there is no other solution it should be properly ordered". Apparently, the purpose of different comms was not to use them.

File: torch/nn/parallel/distributed.py (hunk 12 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: We need to, new commit added the logic

**Comment by teng-li** [comment]
@apaszke  The latest commit has been tested with ResNet50 with the correct 76 accuracy on two nodes. It should be safe to land

File: torch/nn/parallel/distributed.py (hunk 12 above)
Comment (apaszke) [reply torch/nn/parallel/distributed.py]: Do we do that in the other path? I'm not sure now

**Comment by apaszke** [comment]
@teng-li can you please fix the conflict?

**Comment by teng-li** [comment]
@apaszke fixed

File: torch/nn/parallel/distributed.py (hunk 12 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: Yes we do, in here: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L287

**Review by apaszke — APPROVED** [review]
